### ISSUE TITLE
The request is invalid: spec.template.spec.GPUs: GPU nvidia.com/TU104GL_Tesla_T4 is not permitted in permittedHostDevices configuration

### ISSUE THREAD

**Issue by rupang790** [kubevirt/hyperconverged-cluster-operator#1227]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug

**What happened**:
Installed kubevirt-gpu-device-plugin to use GPU on kubevirt VMs on Native K8s (v1.20.5) cluster.
After I edit configmap to add GPU featureGates and CR to add permittedHostDevices information and create VM.yaml, error keeps occurring as below:
```
[root@k8s-master k8s-gpu]# kubectl create -f vm-gpu.yaml
The request is invalid: spec.template.spec.GPUs: GPU nvidia.com/TU104GL_Tesla_T4 is not permitted in permittedHostDevices configuration
```

**What you expected to happen**:
Generate VMs with GPU.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- HCO version (use `oc get csv -n kubevirt-hyperconverged`): 1.4.0
- Kubernetes version (use `kubectl version`): v1.20.5
- Cloud provider or hardware configuration: Hardware
- Install tools:
- Others:

I saw the similar issues on https://github.com/NVIDIA/kubevirt-gpu-device-plugin/issues/17#issue-799163896 # and I tried to solve it on OKD Cluster. At that comment, it is related with config collision between configmap and CR. So put all config on kubevirt-config CM, but it shows ```Invalid Value```. Please check my comment on that issue about detail situation as:
https://github.com/NVIDIA/kubevirt-gpu-device-plugin/issues/17#issuecomment-813851162

**Comment by nunnatsa** [comment]
The kubevirt-config is deprecated and is not supported anymore. Please see the HCO API documentation: https://github.com/kubevirt/hyperconverged-cluster-operator/blob/master/docs/cluster-configuration.md

**Comment by erkanerol** [comment]
@rupang790 Thanks for reporting your problem. This seems an issue we need to follow with https://github.com/NVIDIA/kubevirt-gpu-device-plugin/issues/17. 

Basically, `HyperConverged` object is the highest level object you can use to configure your cluster for virtualization. By using the CR, we are creating a `Kubevirt` object and other objects. When virt-operator doesn't allow us to create a Kubevirt object generated from the HyperConverged object you provide, the error message is propagated and displayed to you. The error you hit comes from the virt-operator side. 

As @nunnatsa  said, we recommend you read the cluster configuration api and then set the permittedHostDevices as explained there without configuring the configmap. If it works, there is no issue. If it fails with the same error, I recommend you create a Kubevirt object directly with the `permittedHostDevices` configuration like in https://github.com/NVIDIA/kubevirt-gpu-device-plugin/issues/17.  If you hit the same issue exactly,  it means that it is a problem in lower layer like gpu-device-plugin or virt-operator. Otherwise, please ping us again with the resources you use in your tests.

**Comment by rupang790** [comment]
@nunnatsa, at my first attempt, I tried as documentation that you linked, but ```* spec.template.spec.GPUs: GPU nvidia.com/TU104GL_Tesla_T4 is not permitted in permittedHostDevices configuration``` error kept occurring. 

@erkanerol As your recommend, I deleted featureGate from kubevirt-config CM and edit cr as linked documentation but error message was same as before. Also trying to create new CR as https://github.com/NVIDIA/kubevirt-gpu-device-plugin/issues/17#issue-799163896 was failed as below error messages:

```
[root@k8s-master k8s-gpu]# kubectl create -f vm-gpu.yaml
The request is invalid:
* spec.template.spec.GPUs: GPU feature gate is not enabled in kubevirt-config
* spec.template.spec.GPUs: GPU nvidia.com/TU104GL_Tesla_T4 is not permitted in permittedHostDevices configuration
```

So at this time, I tried to add featureGates 'GPU' on hco.cr.yaml (   featureGates: 'GPU' ) and install kubevirt-hco again, but error as blow:
```
Error from server (InternalError): error when creating "hco.cr.yaml": Internal error occurred: failed calling webhook "validate-hco.kubevirt.io": Post "https://hyperconverged-cluster-webhook-service.kubevirt-hyperconverged.svc:4343/validate-hco-kubevirt-io-v1beta1-hyperconverged?timeout=30s": service "hyperconverged-cluster-webhook-service" not found
```

To sum it up,

- No changes on kubevirt-config / add permittedHostDevices on hco.cr -> Fail
- Add 'GPU' at featureGates on kubevirt-config / add permittedHostDevices on hco.cr -> Fail
- Delete all featureGates on kubevirt-config / add permittedHostDevices on hco.cr -> Fail
- Delete all featureGates on kubevirt-config / Create new CR as ```kubevirt.cr``` and add permittedHostDevices -> Fail

Is it error with Nvidia/kubevirt-gpu-device-plugin?

**Comment by nunnatsa** [comment]
What version of the hyperconverged-cluster are you using? how do you install it?

**Comment by rupang790** [comment]
@nunnatsa How I installed HCO is..
1. git clone https://github.com/kubevirt/hyperconverged-cluster-operator.git
2. ./deploy/deploy.sh

When I clone from git, branch was master branch and version of HCO is 1.4.0.

**Comment by nunnatsa** [comment]
Reopen the issue. I think the issue is that the deployment manifest points to an old version of the operator and the webhook images.

**Comment by rupang790** [comment]
@nunnatsa I found the 1.4.0-unstable version of operator on quay.io. Is it fine that if I change image on the operator.yaml , which I cloned from git, and ./deploy.sh for test?

**Comment by nunnatsa** [comment]
Yes. This is the latest build from the master branch. I'm working on a PR to modify the files to point to this image.

**Comment by erkanerol** [comment]
The image references may be old on the master branch. It is ok to fix it. However I think
- The original error comes from the kubevirt side as we talked above and should be tracked there.
- The new error ("hyperconverged-cluster-webhook-service" not found) is an unrelated problem. I just checked head of master and I didn't observe such an issue.

@rupang790 After deploying `deploy/deploy.sh`, you need to wait until everything is OK.  Ensure that the commands below return such outputs. Otherwise, please ping us.

```
$ oc -n kubevirt-hyperconverged get pods -l app.kubernetes.io/component=deployment
NAME                                              READY   STATUS    RESTARTS   AGE
hyperconverged-cluster-operator-b967d4f4c-fqfvd   1/1     Running   0          12m
hyperconverged-cluster-webhook-97d48c9dd-ttddt    1/1     Running   0          12m
```
> See that both of them up&running

```
$ oc -n kubevirt-hyperconverged get service hyperconverged-cluster-webhook-service
NAME                                     TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)    AGE
hyperconverged-cluster-webhook-service   ClusterIP   10.99.89.90   <none>        4343/TCP   12m
```
> See that the service exists
